From 39d742bd37b66532ae6e1916646d82f80129424c Mon Sep 17 00:00:00 2001
From: Yurii Rashkovskii <me@yrashk.com>
Date: Fri, 8 Jun 2018 15:09:37 -0700
Subject: [PATCH] Problem: need a Mac to build SIT on OS X

It's not always possible to have an extra computer
handy.

Solution: provide a docker container to cross-compile
and make it easy to use through `make osx` which will
build a release for OS X.
---
 .gitignore                               |  1 +
 Makefile                                 |  7 +++++++
 build-tools/cross-compile-osx/.gitignore |  1 +
 build-tools/cross-compile-osx/Dockerfile | 14 ++++++++++++++
 build-tools/cross-compile-osx/README.md  |  5 +++++
 5 files changed, 28 insertions(+)
 create mode 100644 Makefile
 create mode 100644 build-tools/cross-compile-osx/.gitignore
 create mode 100644 build-tools/cross-compile-osx/Dockerfile
 create mode 100644 build-tools/cross-compile-osx/README.md

diff --git a/.gitignore b/.gitignore
index f786eaf..561a1c4 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
 target
 .idea
 .sit/items/*/.lock
+._docker_osx
diff --git a/Makefile b/Makefile
new file mode 100644
index 0000000..0aefe4b
--- /dev/null
+++ b/Makefile
@@ -0,0 +1,7 @@
+osx: target/x86_64-apple-darwin/release/sit target/x86_64-apple-darwin/release/sit-web
+
+target/x86_64-apple-darwin/release/sit target/x86_64-apple-darwin/release/sit-web: build-tools/cross-compile-osx/Dockerfile sit-core sit-web sit
+	docker build --iidfile ._docker_osx build-tools/cross-compile-osx
+	sed -i s/sha256://g ._docker_osx
+	docker run -u `id -u`:`id -g` -v `pwd`:/sit -w /sit -t `cat ._docker_osx` sh -c "cargo build --release --target=x86_64-apple-darwin"
+	rm -f ._docker_osx
diff --git a/build-tools/cross-compile-osx/.gitignore b/build-tools/cross-compile-osx/.gitignore
new file mode 100644
index 0000000..5b5667a
--- /dev/null
+++ b/build-tools/cross-compile-osx/.gitignore
@@ -0,0 +1 @@
+MacOSX10.11.sdk.tar.xz
diff --git a/build-tools/cross-compile-osx/Dockerfile b/build-tools/cross-compile-osx/Dockerfile
new file mode 100644
index 0000000..e149781
--- /dev/null
+++ b/build-tools/cross-compile-osx/Dockerfile
@@ -0,0 +1,14 @@
+FROM rust:1.26.2
+RUN apt-get update && apt-get install -y clang autotools-dev automake cmake libfuse-dev fuse git
+RUN rustup target add x86_64-apple-darwin
+RUN git clone https://github.com/tpoechtrager/osxcross && cd osxcross && git checkout 1a1733a
+COPY MacOSX10.11.sdk.tar.xz /osxcross/tarballs/
+RUN cd osxcross && UNATTENDED=1 OSX_VERSION_MIN=10.7 ./build.sh
+RUN ln -s /osxcross/target/bin/x86_64-apple-darwin15-ld  /osxcross/target/bin/ld
+ENV PATH="/osxcross/target/bin:$PATH"
+ENV CARGO_TARGET_X86_64_APPLE_DARWIN_LINKER x86_64-apple-darwin15-clang
+ENV CC_X86_64_APPLE_DARWIN_LINKER x86_64-apple-darwin15-clang
+ENV CC x86_64-apple-darwin15-clang
+ENV CMAKE_LINKER x86_64-apple-darwin15-ld
+ENV CMAKE_C_LINK_EXECUTABLE x86_64-apple-darwin15-ld
+ENV CMAKE_C_COMPILER_EXTERNAL_TOOLCHAIN x86_64-apple-darwin15-clang
diff --git a/build-tools/cross-compile-osx/README.md b/build-tools/cross-compile-osx/README.md
new file mode 100644
index 0000000..bba6941
--- /dev/null
+++ b/build-tools/cross-compile-osx/README.md
@@ -0,0 +1,5 @@
+This docker container allows to cross-compile MacOS X build using
+a Docker container. The SDK required is not distributed here but
+can be [extracted from XCode 7.3.1](https://github.com/tpoechtrager/osxcross#packaging-the-sdk)
+or found elsewhere. The only file required is: MacOSX10.11.sdk.tar.xz, its
+SHA256 should be 522c0fd491c7fac4a8660b518c36c9bea8e37f36b10a292eed9a0b57c3dd5604
-- 
2.16.4

